### ISSUE TITLE
Git hook runs on omitted files

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#150]
In GitLab by @cdriehuys on Aug 24, 2016, 15:59

**Python Version:** `Python 3.5.2`

**Installation Method:** `pip`
- *pip version:* `pip 1.8.2`

**setuptools Version:** `25.1.6`

**flake8 Version:** `3.0.4 (pyflakes: 1.2.3, mccabe: 0.5.2, pycodestyle: 2.0.0) CPython 3.5.2 on Linux`

### Problem:

I'm developing a Django project, and I have the following `setup.cfg` file.

```
[flake8]
exclude =
    env/,
    migrations/,

    # exclude tox so flake8 doesn't explode
    .tox/
```

I have also installed the git hook using `flake8 --install-hook git`, which created the following hook.

```python
#!/home/<username>/programs/python/django/django-timetracker/env/bin/python3
import os
import sys

from flake8.main import git

if __name__ == '__main__':
    sys.exit(
        git.hook(
            strict=git.config_for('strict'),
            lazy=git.config_for('lazy'),
        )
    )
```

The issue I'm having is that when I try to commit code in the `migrations` folder that has a line longer than 79 characters, flake8 complains and cancels the commit. However, if I run `flake8 .` from the command line, no errors are reported.

I have tried to delete and reinstall the hook, I've tried various patterns for the exclusion like `*/migrations/*`, `migrations/*`, `timetracker/migrations/`, etc., and I've tried removing the file from git's staging and adding it back in. It still throws a line length error.

**Comment by asottile** [comment]
In GitLab by @asottile on Aug 24, 2016, 16:21

A bit of a shameless plug, but we've run into the same thing while using http://pre-commit.com (a framework for running git hooks) and have opted to instead configure flake8 in the pre-commit configuration (using the slightly-more-powerful `exclude:` configuration option) instead of using flake8's configuration.

The core issue here is `flake8 x.py` will run flake8 on x.py even if x.py is excluded.  Last time I brought this up it was closed as "working as intended".

**Comment by asottile** [comment]
In GitLab by @mikenerone on Aug 25, 2016, 01:14

Actually,  I think the only issue here is that the hook doesn't copy your setup.cfg to the temp directory in which the staged files are scanned. Same issue as #210.

**Comment by asottile** [comment]
In GitLab by @cdriehuys on Aug 25, 2016, 11:35

I don't think that's true, because even if `setup.cfg` is staged, the hook still checks my `migrations` folder, even though `flake8 .` doesn't come up with any errors.

```
$ git status

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   setup.cfg
	modified:   timetracker/migrations/0001_initial.py

$ git commit

/tmp/tmpxnbd6yn0/home/<username>/programs/python/django/django-timetracker/timetracker/migrations/0001_initial.py:23:80: E501 line too long (114 > 79 characters)
```

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Aug 27, 2016, 09:49

Right, I think this is related to #210 (if not exactly the same problem). I think we might be using the wrong "parent" directory for usage with the exclude patterns causing them to be broken for the git-hook.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Aug 27, 2016, 10:10

@asottile you said,

> The core issue here is flake8 x.py will run flake8 on x.py even if x.py is excluded. Last time I brought this up it was closed as "working as intended".

When did you test this? Flake8 3.x behaves as you expect and, I believe, so did 2.x. @The-Compiler can correct me if I'm wrong, but he found that we had regressed that behaviour in the Flake8 3 beta releases.

**Comment by asottile** [comment]
In GitLab by @asottile on Aug 27, 2016, 10:38

It was a very long time ago, essentially this:

### tox.ini
```ini
[flake8]
exclude = vendor
```

### vendor/bad.py
```python
import os
```

### Result
```bash
$ flake8 vendor/bad.py
vendor/bad.py:1:1: F401 'os' imported but unused
```

I found the issue:
- https://gitlab.com/pycqa/flake8/issues/10
- https://github.com/pre-commit/pre-commit-hooks/issues/112 (the issue on my project)

I'm probably incorrect on it being the root cause here, just seemed like the same symptoms :)

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Aug 27, 2016, 10:50

Yeah, I just tested it out. Flake8 has `tests/fixtures` ignored and I added a file with the same thing ('import os') and ran `flake8 tests/fixtures/foo.py` versus `flake8 foo.py`. The former doesn't return anything as being out of place.

**Comment by asottile** [comment]
In GitLab by @stavros on Aug 29, 2016, 06:58

I am also experiencing this. It's very annoying, as I can't commit because my `setup.cfg` is getting ignored.

**Comment by asottile** [comment]
In GitLab by @wjanssen on Sep 27, 2016, 10:31

Same here, as Stavros reports.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Nov 12, 2016, 11:55

This is a duplicate of #210 and fixed on master. It will be released in 3.1.0.
